Makefile

```diff
@@ -1,38 +1,85 @@
-CC=gcc
-CXX=g++
-NVCC=nvcc
+# Load site-specific setting -- controlled using environment variable `site`:
+# eg.  make site=nersc_cori
+ifdef site
+    $(info detected site: $(site))
+    -include sites/make.inc.$(site)
+endif
+
+# Load architecture-specific settings -- controlled using the environment
+# variable `target`: eg. make target=power9
+ifdef target
+    $(info detected target: $(target))
+    -include targets/make.inc.$(target)
+endif
+
+
+CC   ?= gcc
+CXX  ?= g++
+NVCC ?= nvcc
 
 # Developer-users are suggested to optimize NVARCH in their own make.inc, see:
 #   http://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
-NVARCH = -arch=sm_70 \
-	-gencode=arch=compute_35,code=sm_35 \
-	-gencode=arch=compute_50,code=sm_50 \
-	-gencode=arch=compute_52,code=sm_52 \
-	-gencode=arch=compute_60,code=sm_60 \
-	-gencode=arch=compute_61,code=sm_61 \
-	-gencode=arch=compute_70,code=sm_70 \
-	-gencode=arch=compute_75,code=sm_75 \
-	-gencode=arch=compute_75,code=compute_75
-
-CFLAGS= -fPIC -O3 -funroll-loops -march=native -g
-CXXFLAGS= $(CFLAGS) -std=c++14
-NVCCFLAGS= -std=c++14 -ccbin=$(CXX) -O3 $(NVARCH) -Wno-deprecated-gpu-targets \
-	--default-stream per-thread -Xcompiler "$(CXXFLAGS)"
+NVARCH ?= -arch=sm_70 \
+	  -gencode=arch=compute_35,code=sm_35 \
+	  -gencode=arch=compute_50,code=sm_50 \
+	  -gencode=arch=compute_52,code=sm_52 \
+	  -gencode=arch=compute_60,code=sm_60 \
+	  -gencode=arch=compute_61,code=sm_61 \
+	  -gencode=arch=compute_70,code=sm_70 \
+	  -gencode=arch=compute_75,code=sm_75 \
+	  -gencode=arch=compute_75,code=compute_75
+
+CFLAGS    ?= -fPIC -O3 -funroll-loops -march=native -g
+CXXFLAGS  ?= $(CFLAGS) -std=c++14
+NVCCFLAGS ?= -std=c++14 -ccbin=$(CXX) -O3 $(NVARCH) -Wno-deprecated-gpu-targets \
+	     --default-stream per-thread -Xcompiler "$(CXXFLAGS)"
 
 # For debugging, tell nvcc to add symbols to host and device code respectively,
 #NVCCFLAGS+= -g -G
 # and enable cufinufft internal flags.
 #NVCCFLAGS+= -DINFO -DDEBUG -DRESULT -DTIME
 
-# CUDA Related build dependencies
-CUDA_ROOT=/usr/local/cuda
-INC=-I$(CUDA_ROOT)/include \
-	-Icontrib/cuda_samples
-NVCC_LIBS_PATH=-L$(CUDA_ROOT)/lib64
+# CUDA Related build dependencies -- the user can overwrite CUDA_ROOT using the
+# CUDA_DIR environment variable. If neither (CUDA_ROOT, nor CUDA_DIR) is set,
+# CUDA_ROOT defaults to /usr/local/cuda
+ifeq ($(CUDA_DIR),)
+    CUDA_ROOT ?= /usr/local/cuda
+else
+    CUDA_ROOT := $(CUDA_DIR)
+endif
+
+# Common includes
+INC=-I$(CUDA_ROOT)/include -Icontrib/cuda_samples
+ifdef FFTW_INC
+    $(info detected FFTW_INC -- setting FFTW include directory)
+    INC += -I$(FFTW_INC)
+endif
 
 FFTWNAME=fftw3
 
-LIBS=-lm -lcudart -lstdc++ -lnvToolsExt -lcufft -lcuda -l$(FFTWNAME) -l$(FFTWNAME)f
+# Common libs
+LIBS=-lm -lcudart -lstdc++ -lnvToolsExt -lcufft -lcuda -l$(FFTWNAME)
+
+# Figure out the specifics of the FFTW install
```

Comment:
Why can't we just require the FFTW3 repo, which includes single-prec fftw3f libs? Are there environments with libfftw3 but not libfftw3f ?

Reply:
The `fftw` package on Cori GPU doe not include the single-prec `libfftwf*` libraries. Since we're not explicitly dependent on the single-prec libraries, and since it's possible to build fftw without them why force to user to include them?

Reply:
Hey guys, their inclusion here is just oversight on my part.  This code makes use of `cuFFT`...  FINUFFT makes use of `fftw`.  I think the inclusion here was just oversight in the process of bring them closer together (I was altering makefiles for both). There was a time where there was more dependence on internal structures, but this is not the case now.

I don't think we need fftw shared library at all for cufinufft. This was a good catch. I will take care of removing it in another PR.  For now I think you may proceed with this to facilitate the rest of these changes.  I expect the extra lines would be removed... OR

If you want to wait I can put a PR removing FFTW tomorrow, plus review the time. (Not sure how rushed you are to close this out).

Reply:
Once #46 is merged, and brought in here, the line/changes for this conversation should go away.

Reply:
Gotcha -- I'll wait for #46

Reply:
FFTW should be removed master now. Hope that helps. Thanks

Reply:
Thanks, I've merged the changes from #46 into this PR